client/src/auth.rs

```diff
@@ -0,0 +1,96 @@
+use oauth2::{
+    basic::{BasicClient, BasicTokenResponse},
+    url::ParseError,
+    AuthUrl, ClientId, ClientSecret, TokenResponse, TokenUrl,
+};
+use thiserror::Error;
+use tonic::{metadata::MetadataValue, service::Interceptor};
+use tracing::instrument;
+
+#[derive(Debug)]
+pub struct OAuth2Config {
+    pub client_id: String,
+    pub client_secret: String,
+    pub auth_server: String,
+    pub audience: String,
+}
+
+#[derive(Debug)]
+pub struct OAuth2Provider {
+    client: BasicClient,
+    config: OAuth2Config,
+}
+
+#[derive(Error, Debug)]
+pub enum AuthError {
+    #[error("Token request failed")]
+    TokenRequestFailed,
+}
+
+impl OAuth2Provider {
+    fn from_config(config: OAuth2Config) -> Result<OAuth2Provider, ParseError> {
+        let client = BasicClient::new(
+            ClientId::new(config.client_id.clone()),
+            Some(ClientSecret::new(config.client_secret.clone())),
```

Comment:
:question: Cloned or to_owned?

Reply:
I prefer `to_owned` because it describes the intent but again, personal preference: https://users.rust-lang.org/t/what-is-the-difference-between-clone-and-to-owned/54705